### ISSUE TITLE
New Backend: nats.io - Implement a Remote backup Repositories

### ISSUE THREAD

**Issue by Fishwaldo** [restic/restic#3545]
Hi,
I'm working on a project for a client that has deployed Restic on over > 300 clients/Servers. 

They recently had a security issue where a user was able to grab both their B2 credentials as well as the restic password from the environment variables at `/proc/<pid>/environ` (on linux) and cause some mayhem:
1) They restored backups from other clients and potentially obtained some sensitive information (passwords etc) from the backups
2) The abused the B2 buckets credentials they obtained to upload arbitrary files (not related to Restic) and share it publicly. 

They have also had a few other issues with this size deployment:
1) Getting a window for Check/Prune etc is proving to be very difficult. Almost 24x7 there is some client/server that has a backup in progress. Right now, they "hack" around this by temporary changing the credentials on their B2 buckets to block new clients starting backups etc
2) Thundering Hurd/Bandwidth issues when backups are scheduled to kickoff on clients (they randomize it on each client within a 1 hour window, but that's not helping a lot)

We looked at a few options with them to address the issues (primarily to address the credential leaks) and settled upon implementing a new "back-end" that uses [nats.io](https://nats.io/) as a transport and implement "workers" on the other side that does the actual reading/writing to the repositories (B2 and FS back end currently). This has been running as a proof of concept for the last month or so, and they have now requested me to do a full implementation and upstream it to restic!

I'm currently refactoring the restic changes (its currently a "hack" to get it working) and getting the new "workers" into shape as well, but thought I'd open a issue here to gauge the sentiment that you guys would take this as a PR when ready.

Some of the features:
1) Restic now authenticates to Nats.io and not to the back-ends (B2 etc) anymore
2) The "workers" communicate with the back-ends and dynamically map the clients to the correct back-end. This way, you can have a single "configuration" of restic for all devices, and the workers decide which "repository" to actually save the data into.
  eg: `restic -r nats://nats.example.com/<backup location>` - > on the worker, we can map `<backup location>` to the actual B2/FS repository dynamically, hence the clients don't actually know where they are backing up to. It also enables administrators to change these locations without having to touch the clients. 

3) They would like to restrict "standard" restic accounts from reading the /data/ directory (and provide alternative temporary credentials  for restore/check etc operations - (This I've not investigated at all yet, it might be impossible - feedback is welcome)
3) The workers are "stateless" - meaning to scale up, you just need to spin up more workers. This also works very well for my client, as they have 3 main sites (with Nats "Clusters") , plus a number of remote clients. By configuring nats clustering and replication etc, they can place workers at each site to handle local clients rather than having to replicate the data across their WAN links. 
4) Idempotent caching of files such as config, locks, keys, index, snapshots and directory listing across workers (as long as no network partitioning events) 
5) Concurrency control (a bit of a hack) - We can limit the number of clients that "Open" a repository, returning a error to those clients that exceed the limit. Using a "retry" wrapper on the clients, they will retry the backup at a later time. (would love to get access to the back-off logic in the back ends in a sane way to do this in-process on the restic binary

As mentioned, I'm cleaning up both the changes for Restic as well as a "worker" that can either deployed as a "all-in-one" meaning no need to deploy Nats infrastructure (or if the size warrants it, a "slim" worker that doesn't have integrated nats server that can use existing nats infrastructure) 

if this would likely to be accepted, I'll probably open a PR on restic in the next few weeks (but you can see the current progress at https://github.com/Fishwaldo/restic/tree/natsio and the "workers" at https://github.com/Fishwaldo/restic-nats-server/)

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Well I got a small programming gig out of restic as the client wants to continue to use it moving forward, so I would say, yes, I"m happy !

**Comment by smlx** [comment]
Why can't you pass individual client credentials directly to restic when it is backing up each client?

**Comment by smlx** [comment]
> By client credentials are you referring to the actual S3/B2 etc login credentials?
> 
> If so, that relates to point no 1 in my justification above - My client had a rouge user that stole the B2 credentials and repository password from the installation of restic on their machine and did nefarious things with it. So this is designed that no credentials (other than what is needed to login to the Nats server) is required on the client side.

What I mean is can you give each client its own bucket, S3 credentials, and repo password?

**Comment by Fishwaldo** [comment]
(preface - Some of these decisions are driven by my clients IT Security Group - I don't necessarily agree with how much it reduces the risk, but do understand their concerns)

>     * What is the benefit of using nats instead of the existing rest protocol? Using an HTTP load-balancer it should also be possible to scale that up. 
> 

* Guaranteed delivery (built on top of the Nats Protocol, as well as utilizing some of the Nats features) , queuing of "requests" for workers to process (can limit thundering hurd issues when a lot of clients kick off). Localized Processing of requests (so site a client from Site 2 visits site 1, then the workers at Site 1 would handle the client, not the "default" site2 workers. 
* Nats Security/Permissions feature - able to Partition Clients, Decentralized Authentication (using something akin to JWT Claims) 
* (Clients Assessment) using the Rest/HTTP protocol - Doesn't really address the abuse of the B2 buckets. Its basically acting as a "proxy" to the repository and they felt that it made it more insecure as now standard tools such as curl/wget could be used to abuse the storage (where as the user that triggered this installed the B2 command line apps to do the abuse. Nats helps as the security permissions/controls can be applied at the Nats Server/and Worker level and we can leverage Nats distributed authentication to expire/revoke/alter authentication/authorization. 
* (Client Specific) Leverage existing infrastructure in place already for High Availability/Scaling (only need to provision Workers, not Load Balancers etc)

>We're not too keen on introducing additional back-ends as these increase the maintenance work.

I'm on the hook to support this for at least 1 year! (regardless if its up-streamed or not)

>     * To just hide the backend credentials from the restic client, it would also be possible to use `rclone serve restic` running at a server acting as proxy between restic and the backend. How does this compare to your implementation?
> 

We looked at this - two things that came back - 
* HTTP Protocol (as above). 
* They wanted the ability to dynamically maintain the mapping between restic clients and actual repositories. With the current nats platform, all clients can have identical server urls's (-r nats://host/bucket) but on the worker we can dynamically map the clients to different repositories based on username/hostname etc. Right now, they have essentially 3 buckets. Servers, Users and VIP (VIP was only introduced after the security breach above). They would like to be able to manage this mapping centrally without having to touch client installations (altering the -r on each device)

> 
> [Edit]After thinking about this a bit more, what about implementing this as a client-side proxy which serves as a replacement for `rclone serve restic`? That proxy would have to speak using the rest protocol over HTTP/2 via stdin/out with restic. Then it could be used as `restic backup -o rclone.program=nats-proxy -o rclone.args="serve" -r rclone: /path/to/backup`[/Edit]
> 

I'll have a look into that. It may work. As mentioned previously, it would be great to be able hook into the retrybackend a bit more to be able to return "custom" retry times for a failed operation. 


> > Getting a window for Check/Prune etc is proving to be very difficult. Almost 24x7 there is some client/server that has a backup in progress. Right now, they "hack" around this by temporary changing the credentials on their B2 buckets to block new clients starting backups etc
> 
> Are we talking of a single restic repository here that's shared between all clients? Or does each client have its only repository, which just happens to be stored in the same B2 bucket?
> 

As mentioned above, 3 repositories. A initial big draw from them was the 0 configuration maintenance needed on clients after installation of restic, but thats now a drawback for them. 


> > They would like to restrict "standard" restic accounts from reading the /data/ directory (and provide alternative temporary credentials for restore/check etc operations - (This I've not investigated at all yet, it might be impossible - feedback is welcome)
> 
> `backup --force` might be able to work without access to the data directory. However, that has the huge downside that restic has to read everything again. So that's not really a solution. It sounds like you're looking for the asymmetric encryption part of #187, but that is unlikely to manifest in the foreseeable future.

I suspected as much. The Alternative for them will probably to log/audit /data access on the workers and trigger a alert if a client exceeds a threshold. Is there any determinable pattern how /data is accessed during backup compared to a restore. eg, like we only a reading the headers of a pack, versus the whole file, or should I just experiment with a x number of files read in y seconds type thing?
